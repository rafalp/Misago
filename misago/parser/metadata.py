from typing import Iterable

from django.conf import settings
from django.contrib.auth import get_user_model

from ..core.utils import slugify
from .context import ParserContext
from .hooks import (
    get_ast_metadata_users_queryset_hook,
    update_ast_metadata_hook,
    update_ast_metadata_from_node_hook,
    update_ast_metadata_users_hook,
)

User = get_user_model()


def create_ast_metadata(
    context: ParserContext,
    ast: list[dict],
) -> dict:
    metadata = {
        "outbound-links": set(),
        "usernames": set(),
        "users": {},
        "posts": {
            "ids": set(),
            "objs": {},
        },
    }

    return update_ast_metadata_hook(_update_ast_metadata_action, context, ast, metadata)


def _update_ast_metadata_action(
    context: ParserContext,
    ast: list[dict],
    metadata: dict,
) -> dict:
    for ast_node in ast:
        update_ast_metadata_from_node(context, ast_node, metadata)

    update_ast_metadata_users(context, metadata)

    return metadata


def update_ast_metadata_from_node(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
) -> None:
    update_ast_metadata_from_node_hook(
        _update_ast_metadata_from_node_action, context, ast_node, metadata
    )


def _update_ast_metadata_from_node_action(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
) -> None:
    if ast_node["type"] == "mention":
        metadata["usernames"].add(slugify(ast_node["username"]))

    elif ast_node["type"] == "quote-bbcode":
        if ast_node["author"]:
            metadata["usernames"].add(slugify(ast_node["author"]))
        if ast_node["post"]:
            metadata["posts"]["ids"].add(ast_node["post"])

    elif ast_node["type"] in ("auto-link", "auto-url", "url", "url-bbcode"):
        if not context.forum_address.is_inbound_link(ast_node["href"]):
            metadata["outbound-links"].add(ast_node["href"])

    if ast_node.get("children"):
        for child_node in ast_node["children"]:
            update_ast_metadata_from_node(context, child_node, metadata)

    if ast_node.get("items"):
        for child_node in ast_node["items"]:
            update_ast_metadata_from_node(context, child_node, metadata)

    if ast_node.get("lists"):
        for child_node in ast_node["lists"]:
            update_ast_metadata_from_node(context, child_node, metadata)


def update_ast_metadata_users(context: ParserContext, metadata: dict) -> None:
    update_ast_metadata_users_hook(_update_ast_metadata_users_action, context, metadata)


def _update_ast_metadata_users_action(context: ParserContext, metadata: dict) -> None:
    if not metadata["usernames"]:
        return

    usernames = sorted(metadata["usernames"])
    if len(usernames) > settings.MISAGO_PARSER_MAX_USERS:
        return

    if usernames:
        queryset = get_ast_metadata_users_queryset(context, usernames)
        for user in queryset:
            metadata["users"][user.slug] = user


def get_ast_metadata_users_queryset(
    context: ParserContext, usernames: list[str]
) -> Iterable[User]:
    return get_ast_metadata_users_queryset_hook(
        _get_ast_metadata_users_queryset_action, context, usernames
    )


def _get_ast_metadata_users_queryset_action(
    context: ParserContext, usernames: list[str]
) -> Iterable[User]:
    return User.objects.filter(slug__in=usernames)
