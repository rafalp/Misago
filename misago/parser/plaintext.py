from enum import StrEnum

from ..core.utils import slugify
from .context import ParserContext
from .exceptions import AstError
from .hooks import render_ast_node_to_plaintext_hook
from .urls import clean_href


class PlainTextFormat(StrEnum):
    META_DESCRIPTION = "meta_description"
    SEARCH_DOCUMENT = "search_document"


def render_ast_to_plaintext(
    context: ParserContext,
    ast: list[dict],
    metadata: dict,
    text_format: str | None = None,
) -> str:
    plain_text = []
    for ast_node in ast:
        node_text = render_ast_node_to_plaintext(
            context,
            ast_node,
            metadata,
            text_format,
        ).strip()
        if node_text:
            plain_text.append(node_text)
    return (" ".join(plain_text)).strip()


def render_inline_ast_to_plaintext(
    context: ParserContext,
    ast: list[dict],
    metadata: dict,
    text_format: str | None = None,
) -> str:
    plain_text = ""
    for ast_node in ast:
        plain_text += render_ast_node_to_plaintext(
            context,
            ast_node,
            metadata,
            text_format,
        )
    return plain_text


def render_ast_node_to_plaintext(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
    text_format: str | None,
) -> str:
    return render_ast_node_to_plaintext_hook(
        _render_ast_node_to_plaintext_action,
        context,
        ast_node,
        metadata,
        text_format,
    )


AST_INLINE_NODES = (
    "heading",
    "heading-setex",
    "paragraph",
    "emphasis",
    "emphasis-underscore",
    "strong",
    "strong-underscore",
    "strikethrough",
    "strikethrough-bbcode",
    "bold-bbcode",
    "italics-bbcode",
    "underline-bbcode",
    "superscript-bbcode",
    "subscript-bbcode",
)


def _render_ast_node_to_plaintext_action(
    context: ParserContext,
    ast_node: dict,
    metadata: dict,
    text_format: str | None,
) -> str:
    ast_type = ast_node["type"]

    if ast_type in AST_INLINE_NODES:
        return render_inline_ast_to_plaintext(
            context, ast_node["children"], metadata, text_format
        ).strip()

    if ast_type == "list":
        items: list[str] = []
        for i, ast_item in enumerate(ast_node["items"]):
            if children := render_inline_ast_to_plaintext(
                context, ast_item["children"], metadata, text_format
            ).strip():
                item = ""
                if ast_node["ordered"]:
                    item += f"{i + 1}."
                else:
                    item += ast_node["sign"]

                item += " " + children
                if lists := render_ast_to_plaintext(
                    context, ast_item["lists"], metadata, text_format
                ):
                    item += " " + lists

                items.append(item)

        return " ".join(items)

    if ast_type in ("code", "code-bbcode", "code-indented"):
        code_lines = [
            line.strip() for line in ast_node["code"].splitlines() if line.strip()
        ]
        if syntax := ast_node["syntax"]:
            return f"{syntax}: " + " ".join(code_lines)
        return " ".join(code_lines)

    if ast_type == "code-inline":
        return ast_node["code"]

    if ast_type in ("quote", "quote-bbcode"):
        children = render_ast_to_plaintext(
            context, ast_node["children"], metadata, text_format
        )

        if author := ast_node.get("author"):
            return f"{author}: {children}"

        return children

    if ast_type == "spoiler-bbcode":
        children = render_ast_to_plaintext(
            context, ast_node["children"], metadata, text_format
        )

        if summary := ast_node["summary"]:
            return f"{summary}: {children}"

        return children

    if ast_type in ("thematic-break", "thematic-break-bbcode"):
        return ""

    if ast_type in ("image", "image-bbcode"):
        alt = ast_node["alt"] or ""
        if text_format == PlainTextFormat.META_DESCRIPTION:
            return alt

        src = clean_href(ast_node["src"])
        return f"{alt} {src}".strip()

    if ast_type in ("url", "url-bbcode"):
        href = clean_href(ast_node["href"])

        if children := render_inline_ast_to_plaintext(
            context, ast_node["children"], metadata, text_format
        ).strip():
            if text_format == PlainTextFormat.META_DESCRIPTION:
                return children
            else:
                return f"{children} {href}"

        if text_format == PlainTextFormat.META_DESCRIPTION:
            return ""

        return href

    if ast_type in ("auto-link", "auto-url"):
        if text_format == PlainTextFormat.META_DESCRIPTION:
            return ""

        return clean_href(ast_node["href"])

    if ast_type == "mention":
        username = slugify(ast_node["username"])
        if username not in metadata["users"]:
            return "@" + ast_node["username"]

        user = metadata["users"][username]
        return "@" + user.username

    if ast_type == "escape":
        return ast_node["character"]

    if ast_type in ("thematic-break", "thematic-break-bbcode", "line-break"):
        return ""

    if ast_type == "text":
        return ast_node["text"]

    raise AstError(f"Unknown AST node type: {ast_type}")
