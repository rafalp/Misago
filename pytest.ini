[pytest]
testpaths = misago plugins
python_files = tests.py test_*.py *_tests.py
env =
    MISAGO_DEBUG=off
    MISAGO_TEST=on
    MISAGO_CACHE_URL=dummy://null
    MISAGO_PUBSUB_URL=memory://
filterwarnings =
    ignore::DeprecationWarning