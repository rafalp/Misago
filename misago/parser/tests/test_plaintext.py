import pytest

from ..metadata import create_ast_metadata
from ..plaintext import PlainTextFormat, render_ast_to_plaintext


@pytest.mark.parametrize(
    "markup",
    (
        "# Hello world!",
        "## Hello world!",
        "### Hello world!",
        "#### Hello world!",
        "##### Hello world!",
        "###### Hello world!",
    ),
)
def test_render_ast_to_plaintext_heading(
    markup, parser_context, parse_markup, snapshot
):
    ast = parse_markup(markup)
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


@pytest.mark.parametrize(
    "markup",
    (
        "Hello world!\n=====",
        "Hello world!\n-----",
    ),
)
def test_render_ast_to_plaintext_setex_heading(
    markup, parser_context, parse_markup, snapshot
):
    ast = parse_markup(markup)
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_quote(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        > Hello world!
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_quote_bbcode(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        [quote]Hello world![/quote]
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_quote_bbcode_with_author(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        """
        [quote=Author]Hello world![/quote]
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_unordered_list(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        - Lorem
        - _Ipsum_
        - Dolor
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_ordered_list(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        1. Lorem
        2. _Ipsum_
        3. Dolor
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_unordered_list_with_nested_list(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        """
        - Lorem
        - _Ipsum_
          - Met
          - Elit
        - Dolor
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_spoiler(parser_context, parse_markup, snapshot):
    ast = parse_markup("[spoiler]Hello world![/spoiler]")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_spoiler_with_summary(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("[spoiler=Secret message]Hello world![/spoiler]")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_paragraph(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello world!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_two_paragraphs(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello world!\n\nHow's going?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_thematic_break(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello world!\n- - -\nHow's going?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_thematic_break_bbcode(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello world!\n[hr]\nHow's going?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_code(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        ```
        if random.randint(0, 10) > 4:
            print("Gotcha!")
            return True
        return False
        ```
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_code_with_syntax(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        """
        ```python
        if random.randint(0, 10) > 4:
            print("Gotcha!")
            return True
        return False
        ```
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_code_bbcode(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        [code]
            if random.randint(0, 10) > 4:
                print("Gotcha!")
                return True
            return False
        [/code]
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_code_bbcode_with_syntax(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        """
        [code=python]
            if random.randint(0, 10) > 4:
                print("Gotcha!")
                return True
            return False
        [/code]
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_code_indented(parser_context, parse_markup, snapshot):
    ast = parse_markup(
        """
        Hello:

            if random.randint(0, 10) > 4:
                print("Gotcha!")
                return True
            return False
        """
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_inline_code(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello `world`!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_emphasis_text(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello *world*!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_emphasis_underscore_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello _world_!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_strong_text(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello **world**!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_strikethrough_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello ~~world~~!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_bold_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [b]world[/b]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_italics_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [i]world[/i]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_underline_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [u]world[/u]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_strikethrough_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [s]world[/s]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_superscript_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [sup]world[/sup]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_subscript_bbcode_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup("Hello [sub]world[/sub]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_escaped_characters(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(r"Hello\[hr\]World!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_soft_linebreak(parser_context, parse_markup, snapshot):
    ast = parse_markup("Hello world!\nHow's going?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_image_auto_link(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See the logo: <!https://misago-project.org/img.png>")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_image(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the logo: !(https://misago-project.org/img.png)")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_image_with_alt_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See the logo: ![Alt text](https://misago-project.org/img.png)")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_image_bbcode(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the logo: [img]https://misago-project.org/img.png[/img]")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_image_bbcode_with_alt_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        f"See the logo: [img=https://misago-project.org/img.png]Alt text[/img]"
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_image_auto_link_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See the logo: <!https://misago-project.org/img.png>")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_image_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See the logo: !(https://misago-project.org/img.png)")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_image_with_alt_text_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See the logo: ![Alt text](https://misago-project.org/img.png)")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_image_bbcode_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See the logo: [img]https://misago-project.org/img.png[/img]")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_image_bbcode_with_alt_text_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(
        f"See the logo: [img=https://misago-project.org/img.png]Alt text[/img]"
    )
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_url(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See [*the site*](https://misago-project.org)!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_url_bbcode(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See [url]https://misago-project.org[/url]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_url_bbcode_with_text(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See [url=https://misago-project.org]*the site*[/url]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_auto_link(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the site: <https://misago-project.org>")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_auto_url(parser_context, parse_markup, snapshot):
    ast = parse_markup(f"See the site: https://misago-project.org")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_url_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See [*the site*](https://misago-project.org)!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_url_bbcode_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See [url]https://misago-project.org[/url]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_url_bbcode_with_text_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See [url=https://misago-project.org]*the site*[/url]!")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_auto_link_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See the site: <https://misago-project.org>")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_auto_url_meta_description(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"See the site: https://misago-project.org")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(
        parser_context,
        ast,
        metadata,
        PlainTextFormat.META_DESCRIPTION,
    )


def test_render_ast_to_plaintext_mention(parser_context, parse_markup, snapshot, user):
    user.id = 1000
    user.set_username("MentionedUser")
    user.set_email("mentioned@example.com")
    user.save()

    ast = parse_markup(f"How's going, @{user.username}?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_mention_non_existing_user(
    parser_context, parse_markup, snapshot
):
    ast = parse_markup(f"How's going, @JohnDoe?")
    metadata = create_ast_metadata(parser_context, ast)
    assert snapshot == render_ast_to_plaintext(parser_context, ast, metadata)


def test_render_ast_to_plaintext_for_unsupported_ast_raises_error(parser_context):
    with pytest.raises(ValueError):
        metadata = create_ast_metadata(parser_context, [])
        render_ast_to_plaintext(parser_context, [{"type": "unsupported"}], metadata)


def test_render_ast_to_plaintext_meta_description_removes_new_lines(
    parser_context, parse_markup
):
    ast = parse_markup(f"Lorem ipsum\nDolor met")
    metadata = create_ast_metadata(parser_context, ast)
    assert (
        render_ast_to_plaintext(
            parser_context,
            ast,
            metadata,
            PlainTextFormat.META_DESCRIPTION,
        )
        == "Lorem ipsum Dolor met"
    )


def test_render_ast_to_plaintext_search_document_removes_new_lines(
    parser_context, parse_markup
):
    ast = parse_markup(f"Lorem ipsum\nDolor met")
    metadata = create_ast_metadata(parser_context, ast)
    assert (
        render_ast_to_plaintext(
            parser_context,
            ast,
            metadata,
            PlainTextFormat.SEARCH_DOCUMENT,
        )
        == "Lorem ipsum Dolor met"
    )
