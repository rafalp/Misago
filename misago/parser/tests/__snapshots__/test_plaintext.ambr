# serializer version: 1
# name: test_render_ast_to_plaintext_auto_link
  'See the site: https://misago-project.org'
# ---
# name: test_render_ast_to_plaintext_auto_link_meta_description
  'See the site:'
# ---
# name: test_render_ast_to_plaintext_auto_url
  'See the site: https://misago-project.org'
# ---
# name: test_render_ast_to_plaintext_auto_url_meta_description
  'See the site:'
# ---
# name: test_render_ast_to_plaintext_bold_bbcode_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_code
  'if random.randint(0, 10) > 4: print("Gotcha!") return True return False'
# ---
# name: test_render_ast_to_plaintext_code_bbcode
  'if random.randint(0, 10) > 4: print("Gotcha!") return True return False'
# ---
# name: test_render_ast_to_plaintext_code_bbcode_with_syntax
  'python: if random.randint(0, 10) > 4: print("Gotcha!") return True return False'
# ---
# name: test_render_ast_to_plaintext_code_indented
  'Hello: if random.randint(0, 10) > 4: print("Gotcha!") return True return False'
# ---
# name: test_render_ast_to_plaintext_code_with_syntax
  'python: if random.randint(0, 10) > 4: print("Gotcha!") return True return False'
# ---
# name: test_render_ast_to_plaintext_emphasis_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_emphasis_underscore_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_escaped_characters
  'Hello[hr]World!'
# ---
# name: test_render_ast_to_plaintext_heading[# Hello world!]
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_heading[## Hello world!]
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_heading[### Hello world!]
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_heading[#### Hello world!]
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_heading[##### Hello world!]
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_heading[###### Hello world!]
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_image
  'See the logo: https://misago-project.org/img.png'
# ---
# name: test_render_ast_to_plaintext_image_auto_link
  'See the logo: https://misago-project.org/img.png'
# ---
# name: test_render_ast_to_plaintext_image_auto_link_meta_description
  'See the logo:'
# ---
# name: test_render_ast_to_plaintext_image_bbcode
  'See the logo: https://misago-project.org/img.png'
# ---
# name: test_render_ast_to_plaintext_image_bbcode_meta_description
  'See the logo:'
# ---
# name: test_render_ast_to_plaintext_image_bbcode_with_alt_text
  'See the logo: Alt text https://misago-project.org/img.png'
# ---
# name: test_render_ast_to_plaintext_image_bbcode_with_alt_text_meta_description
  'See the logo: Alt text'
# ---
# name: test_render_ast_to_plaintext_image_meta_description
  'See the logo:'
# ---
# name: test_render_ast_to_plaintext_image_with_alt_text
  'See the logo: Alt text https://misago-project.org/img.png'
# ---
# name: test_render_ast_to_plaintext_image_with_alt_text_meta_description
  'See the logo: Alt text'
# ---
# name: test_render_ast_to_plaintext_inline_code
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_italics_bbcode_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_mention
  "How's going, @MentionedUser?"
# ---
# name: test_render_ast_to_plaintext_mention_non_existing_user
  "How's going, @JohnDoe?"
# ---
# name: test_render_ast_to_plaintext_ordered_list
  '1. Lorem 2. Ipsum 3. Dolor'
# ---
# name: test_render_ast_to_plaintext_paragraph
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_quote
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_quote_bbcode
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_quote_bbcode_with_author
  'Author: Hello world!'
# ---
# name: test_render_ast_to_plaintext_setex_heading[Hello world!\n-----]
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_setex_heading[Hello world!\n=====]
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_soft_linebreak
  "Hello world! How's going?"
# ---
# name: test_render_ast_to_plaintext_spoiler
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_spoiler_with_summary
  'Secret message: Hello world!'
# ---
# name: test_render_ast_to_plaintext_strikethrough_bbcode_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_strikethrough_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_strong_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_subscript_bbcode_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_superscript_bbcode_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_thematic_break
  "Hello world! How's going?"
# ---
# name: test_render_ast_to_plaintext_thematic_break_bbcode
  "Hello world! How's going?"
# ---
# name: test_render_ast_to_plaintext_two_paragraphs
  "Hello world! How's going?"
# ---
# name: test_render_ast_to_plaintext_underline_bbcode_text
  'Hello world!'
# ---
# name: test_render_ast_to_plaintext_unordered_list
  '- Lorem - Ipsum - Dolor'
# ---
# name: test_render_ast_to_plaintext_unordered_list_with_nested_list
  '- Lorem - Ipsum - Met - Elit - Dolor'
# ---
# name: test_render_ast_to_plaintext_url
  'See the site https://misago-project.org!'
# ---
# name: test_render_ast_to_plaintext_url_bbcode
  'See https://misago-project.org!'
# ---
# name: test_render_ast_to_plaintext_url_bbcode_meta_description
  'See !'
# ---
# name: test_render_ast_to_plaintext_url_bbcode_with_text
  'See the site https://misago-project.org!'
# ---
# name: test_render_ast_to_plaintext_url_bbcode_with_text_meta_description
  'See the site!'
# ---
# name: test_render_ast_to_plaintext_url_meta_description
  'See the site!'
# ---
